### ISSUE TITLE
Provide better diagnostics when patching fails

### ISSUE THREAD

**Issue by jdm** [rust-lang/cargo#7161]
There is one particular case I keep hitting that's a papercut for me. Servo uses a mixture of git and crates.io dependencies, and sometimes we end up switching between the two so it's easy to forget the true source of a dependency. When adding a patch, we either need a `[patch.crates-io]` or `[patch."https://github.com/servo/whatever"]`. When I get this wrong, however, all I get is a message from Cargo that says that the patch was ignored. It seems like it should be possible for Cargo to notice when a patch matches the name of an existing dependency, check whether the patch source differs from the dependency's source and provide a better error message in that case.

**Comment by benesch** [comment]
I'm with you 100%—I've been real frustrated by this! But this is arguably a duplicate of #6341?

**Comment by jdm** [comment]
It overlaps, but I think keeping them separate makes sense - this one is particularly focused on the poor diagnostics.

**Comment by weihanglo** [comment]
There are more recent updates from <https://github.com/rust-lang/cargo/issues/13952>. I think we can close this as duplicate?

**Comment by jdm** [comment]
I don't see anything in #13952 that is a duplicate of the issue I described here.

**Comment by weihanglo** [comment]
True. They are related but not exactly the same. Could possibly be solved by the same mechanism.
